Fix routes for vlan interfaces under debian/ubuntu

mnederlof opened the pull request:
### What does this PR do?
Fixes an issue with network.routes state, when the interface name is a vlan interface (for example bond0.1332)

### What issues does this PR fix or reference?
Whenever a file has a . in the filename (for example when adding routes for bond0.1332 interface, it would then be route-bond0.1332) the route file is skipped. By changing the dots in the iface name for underscores, it is properly executed by ifup/ifdown and routes are activated.

### Tests written?
No

salt-jenkins commented:
@mnederlof, thanks for your PR! By analyzing the history of the files in this pull request, we identified @garethgreenaway, @MTecknology and @s0undt3ch to be potential reviewers.

MTecknology commented:
I was scratching my head until I finally realized that ubuntu probably uses source-directory instead of source... makes perfect sense now.

My only concern is the next person that gets to play with networking stuff. Could you update the comment to mention the matching pattern `^[a-zA-Z0-9_-]+$` for source-directory being the cause of the limitation? (it feels kinda misleading to me as it's written now)

Other than that, lgtm.

mnederlof commented:
Documentation / Comment has been updated as @MTecknology requested

MTecknology commented on salt/modules/debian_ip.py:
```diff
@@ -1498,6 +1498,13 @@ def _write_file_routes(iface, data, folder, pattern):
+    # ^[a-zA-Z0-9_-]+$ not all filenames are legitimate.
+    # In order to make routes work for vlan interfaces (e.g. bond0.12)
+    # the dots in the iface name will be replaced by underscores, so it
+    # will be properly executed
```
I'm sorry, but this doesn't seem to help with clarity. What is being executed? Why are filenames not legitimate?

Perhaps...
`The source-directory option in /etc/network/interfaces uses the pattern "^[a-zA-Z0-9_-]+$" to match filenames, which creates problems with vlan-type interfaces (e.g. bond0.12) that may have a dot in them. This replaces dots with underscores in order to maintain compatibility with both source-directory and source (with glob) file inclusions.`

Ref:
`       Similarly, "source-directory" keyword is used to source multiple  files
       at  once,  without  specifying  them individually or using shell globs.
       Additionally, when "source-directory" is used, names of the  files  are
       checked to match the following regular expression: ^[a-zA-Z0-9_-]+$. In
       other words, the names must consist entirely of ASCII upper- and lower-
       case letters, ASCII digits, ASCII underscores, and ASCII minus-hyphens.
       In the directory path, shell wildcards may be used as well.
`

mnederlof replied:
Well, this patch is specifically for routes, not for interface files that are referenced or included via /etc/network/interfaces.

I have not tested the source-directory thing for interface files, but it could be that it needs to be there as well. In our environment all interfaces are configured in /etc/network/interfaces

mnederlof replied:
because a route file for bond0.12 will be stored by default in /etc/network/if-up.d/route-bond0.12 (which is not executed, because it does not match the pattern)

mnederlof replied:
From what i can find online, ifup / ifdown probably uses run-parts to execute a directory.

From man-page run-parts:
>
> run-parts runs all the executable files named within constraints described below, found in 
> directory directory.  Other files and directories are silently ignored.
>
> If neither the --lsbsysinit option nor the --regex option is given then the names must consist
> entirely of ASCII upper- and lower-case letters, ASCII digits, ASCII underscores, and ASCII minus-hyphens.
>
> (^[a-zA-Z0-9_-]+$)

MTecknology replied:
I seem to have mistaken this for interfaces.d/...

mnederlof replied:
no problem, updated the comment nonetheless :)

mnederlof committed:
Fix routes for vlan interfaces under debian/ubuntu

Whenever a file has a . in the filename (for example when adding
routes for bond0.1332 interface, it would then be route-bond0.1332)
the route file is skipped. By changing the dots in the iface name
for underscores, it is properly executed and handled and routes
are also added for vlan interfaces.

Ubuntu runs all files in the /etc/network/if-up.d directory with a
matching pattern like ^[a-zA-Z0-9_-]+$, using run-parts
